[Boost] Fix features/optimizations initialization

dilirity opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* This update fixes features and optimizations not getting their setup_trigger respected.
* All of them had `init`, but they were actually getting initialized on `plugins_loaded`, which is where the main Optimizations class [gets initialized](https://github.com/Automattic/jetpack/blob/c37aca77f5ed00429b428154ecd12ba6a8a88bba/projects/plugins/boost/app/features/optimizations/Optimizations.php#L138).
* To preserve backwards compatibility, their setup_trigger has been updated to `plugins_loaded`. 

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

p1675866389883209-slack-C016BBAFHHS

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1. Install this version of the plugin;
2. Activate features;
3. Make sure everything works as expected.

dilirity committed:
Update feature/optimization initialization to respect setup_trigger

dilirity committed:
changelog

thingalon commented:
It looks like the Boost E2E tests are failing on this PR.

The changes look fairly straightforward, and should be ok. But let's make sure the E2E tests work first.

haqadn commented on projects/plugins/boost/app/features/optimizations/Optimizations.php:
```diff
@@ -103,7 +103,8 @@ public function setup() {
 			}
 
-			$optimization->feature->setup();
+			add_action( $optimization->feature->setup_trigger(), array( $optimization->feature, 'setup' ) );
```
There is an `Automattic\Jetpack_Boost\Lib\Setup` which has this implemented this `add_action`. The advantage of using that over hooking directly is that using the `Setup` class allows you to `remove_action` on any of the instances' methods or to get access to the exact instance which was setup for any other reason.

dilirity replied:
That actually makes sense! Initially when debugging, I wasn't sure why setuppable ~~features~~ optimizations aren't present in the instances list.

I've fixed this with 380724d9219c364e29b5d8f3f6c5d6ea13da11f2.

haqadn commented on projects/plugins/boost/app/features/optimizations/Optimizations.php:
```diff
@@ -103,7 +103,8 @@ public function setup() {
-			$optimization->feature->setup();
+			add_action( $optimization->feature->setup_trigger(), array( $optimization->feature, 'setup' ) );
+
 			$this->register_endpoints( $optimization->feature );
```
This probably also isn't doing anything. It looks like the endpoints are being registered in the features individually. And the `Optimizations::register_endpoints` function doesn't do any work. If I am right about this, perhaps it would be good also remove this call and function implementation as this refactor would be relevant to your PR.

dilirity replied:
Thanks! I hadn't noticed this. I actually updated optimization endpoint registration to happen inside Optimizations->setup as that seems to make the most sense. Done with 65c0bef.

Though on one hand we could remove the Has_Endpoints contract and leave features decide how to register endpoints. On the other - simply declaring a list of endpoints and leaving the main class to handle registration makes the most sense (and saves 1 line per feature 😅).

haqadn reviewed (changes requested):
It looks like, in some of the cases, the rest endpoints aren't being registered. ~~It is not always reproducible~~, and I am not sure what's causing it. Probably because we are adding `plugins_loaded` within `plugins_loaded`. In that case, maybe we should use `init` for the individual optimization features.

Update: You can launch a Jurassic ninja and try to regenerate Critical CSS to reproduce the breakage.

pyronaur commented:
Looks like the E2E tests are breaking

dilirity committed:
Refactor feature/optimization initialization

dilirity committed:
Update optimizations setup to be properly registered

dilirity committed:
Update Optimizations instance to handle feature endpoint initialization

This will update features to no longer take care of rest api
initialization themselves.

dilirity commented:
@Automattic/heart-of-gold  Thanks for the heads-up on the e2e tests guys. I've seen them fail in prior PRs which is probably why I ignored them. I've taken note to double check things for future PRs.

haqadn reviewed (dismissed):
I think this looks good to merge now. 👍🏼 

Though I have one thought. @Automattic/heart-of-gold , what do you think about simplifying it further and just having one `setup` method on all `Has_Setup` classes that only does hooking? The `Setup` class can then just call that instantly. 

The current implementation already has a `setup` method which is very similar. The only difference is that it does some non-hook things. We could move them to their own method(e.g. init) and hook appropriately when necessary. 

The current way of hooking to setup a hook to setup hooking looks like unnecessary steps.

dilirity commented:
@haqadn wouldn't that lead to the same problem we had before, where we didn't have control over where and when we hook the setup to?

I played around with the idea of simplifying, but I couldn't figure out a way to avoid the two methods.

dilirity committed:
Simplify setup process, to avoid unnecessary hooking

dilirity commented:
@haqadn I refactored it to remove the unnecessary hooking in [b9d2c7a](https://github.com/Automattic/jetpack/pull/28888/commits/b9d2c7a0adaff6db56c188cabf0de8f97975b24e).

You're right - it makes more sense now.

haqadn commented on projects/plugins/boost/app/lib/Setup.php:
```diff
@@ -27,7 +27,7 @@ class Setup {
 	 */
 	public static function add( Has_Setup $instance ) {
-		self::$instances[] = $instance->do_setup();
+		self::$instances[] = $instance->setup();
```
`setup` may not be returning the instance.

```suggestion
		$instance->setup()
		self::$instances[] = $instance;
```

dilirity replied:
👀 [884e1c8](https://github.com/Automattic/jetpack/pull/28888/commits/884e1c8403d268d499bb551ca8d54af8be6a01db)

haqadn reviewed (changes requested):
Just a small fix

dilirity committed:
Fix Setup not collecting instances

dilirity committed:
Fix optimizations initializing before plugins_loaded

dilirity commented:
Optimizations should now be initialized on `plugins_loaded`, priority 10. Fixed with https://github.com/Automattic/jetpack/pull/28888/commits/92290fe6ae91c2ff57218c926e8fc19a2492a707.

Props to @haqadn for catching this!

dilirity committed:
Merge branch 'trunk' into fix/boost/features-init

dilirity committed:
Merge branch 'trunk' into fix/boost/features-init

haqadn reviewed (approved):
LGTM!

thingalon reviewed (approved):
Echoing @haqadn 's approval since something broke
